### ISSUE TITLE
Serverless Framework project fails to deploy because of archive size

### ISSUE THREAD

**Issue by NikitaHarzha** [prisma/prisma#6644]
## Bug description
Generated bundle of the Serverless project has too big size and exceeds allowed by AWS Lambda 250 MB.
It happens because of big sizes of downloaded @prisma/client, which, for some reasons, downloads engines folder that has unzipped size of 96 MB. And the other big files are connected to generated using command `prisma generate` schema.
![image](https://user-images.githubusercontent.com/80341254/115226713-3aa34e00-a118-11eb-964e-d6df8582d792.png)
![image](https://user-images.githubusercontent.com/80341254/115226766-498a0080-a118-11eb-90cf-e05758a52bd5.png)
![image](https://user-images.githubusercontent.com/80341254/115226823-573f8600-a118-11eb-9a6b-e8abc6a444d4.png)

## Dependency versions
@prisma/client: "^2.20.0"
prisma (dev) : "^2.20.1"
serverless (dev): "^2.32.1"
serverless-webpack: "^5.4.0"
webpack: "^5.11.0"
@types/webpack: "^4.41.25"

## Webpack config
```import { Configuration } from 'webpack';
import nodeExternals from 'webpack-node-externals';
import slsw from 'serverless-webpack';
import { TsconfigPathsPlugin } from 'tsconfig-paths-webpack-plugin';
import CopyPlugin from 'copy-webpack-plugin';

const isOffline = !!slsw.lib.webpack.isLocal;

if (isOffline) {
    // eslint-disable-next-line no-console
    console.log('Offline mode detected!');
}

// Add plugins for the offline mode here
const offlinePlugins: NonNullable<Configuration['plugins']> = [];

const config: Configuration = {
    target: 'node',
    entry: slsw.lib.entries,
    externals: [nodeExternals()],
    mode: isOffline ? 'development' : 'production',
    devtool: 'source-map',
    module: {
        rules: [
            {
                test: /\.ts$/,
                use: [
                    'source-map-loader',
                    {
                        loader: 'ts-loader',
                        options: {
                            onlyCompileBundledFiles: true,
                        },
                    },
                ],
            },
            {
                test: /\.graphql$/,
                use: ['raw-loader'],
            },
        ],
    },
    plugins: [
        ...(isOffline ? offlinePlugins : []),
        new CopyPlugin({
            patterns: [
                { from: './src/database/prisma/schema.prisma', to: './schema.prisma' },
            ],
        })],
    resolve: {
        plugins: [
            new TsconfigPathsPlugin(),
        ],
        extensions: ['.ts', '.js'],
    },
    optimization: {
        minimize: !isOffline,
    },
    stats: 'minimal',
};

export = config;
```

## Serverless webpack options
```webpack:
    webpackConfig: webpack.config.ts
    includeModules: true
    packager: npm
    packagerOptions:
      scripts:
        - prisma generate
```

## Environment 
& setup
- OS:  Mac OS Big Sur 11.1
- Database: PostgreSQL
- Node.js version: 12
- Prisma version: 2.20.1

## Background
This is on the first deploy of Prisma to our existing Lambda stack. We have not been able to get Lambda working yet. We bundle our functions through webpack before deploying.

I have looked at the webpack bundle and confirm that @prisma/client is being bundled with the lambda function before deploying.

**Comment by janpio** [comment]
Did I understand correctly that `Serverless project` refer to a Serverless Framework project? And you are deploying with `serverless deploy` and then getting an error from AWS Lambda about the archive upload size?

It seems you already added `binaryTargets = ["native", "rhel-openssl-1.0.x"]` to your schema file, is that correct?

The challenge here fundamentally is that your local environment needs a different binary than the deployment platform, AWS Lambda. That is the reason why you have two binaries `query-engine...` and by default it seems both are included in the archive. That is not good and we will figure out a way to configure your project not to do that.

**Comment by NikitaHarzha** [comment]
@janpio Yeah, I am already added `binaryTargets = ["native", "rhel-openssl-1.0.x"]` to a schema file.
I am deploying using `serverless deploy` and package size exceeds 250 MB. That's reason why deploy fails.
![image](https://user-images.githubusercontent.com/80341254/115230247-c1f2c080-a11c-11eb-93ec-63081c53ef7b.png)

**Comment by janpio** [comment]
Thanks for confirming. Can you make sure the above message contains all the information so we can easily and quickly reproduce your Serverless project on our side? I will want to set up a reproduction of your problem and double check what is being uploaded to AWS Lambda etc.

**Comment by NikitaHarzha** [comment]
@janpio 
Here is my `serverless.yaml` file
```
service: prisma-test
useDotenv: true
resources: ${file(resources.yaml)}

plugins:
  - serverless-offline
  - serverless-iam-roles-per-function
  - serverless-webpack

provider:
  name: aws
  runtime: nodejs12.x
  stage: ${opt:stage, 'dev'}
  region: ${opt:region, 'us-east-1'}
  memorySize: 512
  timeout: 30
  versionFunctions: false
  apiGateway:
    shouldStartNameWithService: true
  variableSyntax: "\\${((?!AWS|SecretName)[ ~:a-zA-Z0-9._@'\",\\-\\/\\(\\)\\*\\?]+?)}"
  environment:
    DB_USERNAME: ${self:custom.db.username}
    DB_PASSWORD: ${self:custom.db.password}
    DATABASE_URL: ${self:custom.db.dbUrl}
functions:
  TestFunction:
    handler: src/lambdas/test.handler
    name: ${self:custom.servicePrefix}-test
    events:
      - http:
          path: test
          method: get
          cors: true
      
custom:
  webpack:
    webpackConfig: webpack.config.ts
    includeModules: true
    packager: npm
    packagerOptions:
      scripts:
        - prisma generate
  servicePrefix: ${self:service}-${self:provider.stage}
  db:
    username: ${env:DB_USERNAME}
    password: ${env:DB_PASSWORD}
    dbUrl: ${env:DATABASE_URL}
```
## Project dependencies and scripts
```
{
  "name": "sls",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "sls": "sls -c ./serverless.yaml",
    "build": "npm run sls package",
    "deploy": "npm run sls deploy -- -v",
    "undeploy": "npm run sls remove",
    "start": "npm run sls offline start -- --skipCacheInvalidation",
    "prisma:generate": "prisma generate",
    "prisma:migrate": "prisma  migrate dev --name  init  --preview-feature",
    "tslint": "eslint --ext .ts ./src"
  },
  "prisma": {
    "schema": "./src/database/prisma/schema.prisma"
  },
  "husky": {
    "hooks": {
      "pre-commit": "npm run tslint"
    }
  },
  "author": "LambdaTeam",
  "license": "ISC",
  "devDependencies": {
    "@types/copy-webpack-plugin": "^6.4.1",
    "@types/node": "^14.14.37",
    "@types/uuid": "^8.3.0",
    "@types/webpack": "^4.41.25",
    "@types/webpack-node-externals": "^2.5.1",
    "@typescript-eslint/eslint-plugin": "^4.20.0",
    "aws-sdk": "^2.877.0",
    "copy-webpack-plugin": "^8.1.1",
    "eslint": "^7.23.0",
    "eslint-config-airbnb-typescript": "^12.3.1",
    "eslint-plugin-import": "^2.22.1",
    "husky": "^6.0.0",
    "prisma": "^2.20.1",
    "serverless": "^2.32.1",
    "serverless-iam-roles-per-function": "^3.1.0",
    "serverless-offline": "^6.9.0",
    "serverless-webpack": "^5.4.0",
    "source-map-loader": "^2.0.1",
    "ts-loader": "^8.1.0",
    "ts-node": "^9.1.1",
    "tsconfig-paths-webpack-plugin": "^3.5.1",
    "typescript": "^4.2.3",
    "webpack": "^5.11.0",
    "webpack-node-externals": "^2.5.2"
  },
  "dependencies": {
    "@hapi/boom": "^9.1.2",
    "@prisma/client": "^2.20.1",
    "middy": "^0.36.0",
    "uuid": "^8.3.2"
  }
}

```

**Comment by janpio** [comment]
Quick update: I can confirm that we upload both binaries, the native and the `rhel` one, in standard Serverless Framework projects. No Webpack required even. If your own application code is bug, this can of course lead to upload (50MB limit) or unpack (150MB limit) problems. I'll look for a solution and let you know again.

**Comment by janpio** [comment]
Just a note along the way: You can run `serverless package` to create a local version of the archive that is uploaded, then run `less foo.zip` to get a nice view of the contents that makes it easy to identify what is taking space in there.

**Comment by janpio** [comment]
And here is a quick workaround to unblock you @NikitaHarzha:

Add this block to your `.serverless.yml` file:
```
package:
  exclude:
    - node_modules/.prisma/client/query-engine-darwin
```
This will exclude the superfluous binary file from packaging into your zip archive that is uploaded to AWS Lamda.

**Comment by janpio** [comment]
And here is the generalized version of the configuration:

```
# only include the Prisma binary required on AWS Lambda while packaging
package:
  patterns:
    - '!node_modules/.prisma/client/query-engine-*'
    - 'node_modules/.prisma/client/query-engine-rhel-*'
```

This is pretty explicit: It includes all `query-engine-*` files from `node_modules/.prisma/client/`, but then includes all the `query-engine-rhel-*` ones (which only 1 exists). This way you only get 1 Query Engine binary in your packaged archived that is uploaded to AWS Lambda.

(We are now also testing this in our e2e tests at https://github.com/prisma/e2e-tests/tree/dev/platforms-serverless to make sure it does not break with updates in tooling.)

Can you confirm @NikitaHarzha using this package pattern in your serverless configuration fixes the problem for you?

---

By the way @NikitaHarzha, can you still share the error message you are getting? That would help for people to arrive here via Google when they have the same problem as you had - and then they can find the solution right here as well 👍

**Comment by NikitaHarzha** [comment]
@janpio Thanks for your help!
Unfortunately, it didn't worked for me. Files are still included in a package and package size is too big.
I tried to implement this in different ways, but still I am not able to deploy my changes
![image](https://user-images.githubusercontent.com/80341254/115381115-eeb9dd00-a1db-11eb-80ff-ba8134209d4a.png)

**Comment by janpio** [comment]
Hm, that is weird. Some files _need_ to be included of course, otherwise Prisma can not work - so the `node_modules/.prisma/client/query-engine-rhel-*` file is indeed required and that takes ~50MB unpacked.

1. Did you deploy this project before successfully?
2. Can you use `serverless package` and then `less .sererless/####.zip` to look into the generated archive, what is taking space in there?
	- Feel free to also send me the zip file privately to piotrowski@prisma.io if you want me to take a look. I know what to look for now.
4. Is your project expected to use >200MB for the rest of the code?

**Comment by NikitaHarzha** [comment]
@janpio I deployed this project successfully, but it was deployed without `prisma generate` command and still had 50 MB zipped size.
I've checked archived file and it seems that these files take all the space there
![image](https://user-images.githubusercontent.com/80341254/115384954-478b7480-a1e0-11eb-9bf0-423f3cf033ec.png)

Expected project size is about 5 MB, there is no code and dependencies that can use such a big files, except Prisma

**Comment by janpio** [comment]
Oooookay, seems you are bundling the CLI as well here - I overlooked that before. 

Where do these files live? 
Do you need the CLI in the built project? Otherwise, if you put the `prisma` dependency into the `devDependencies` of your project this would automatically be excluded from the package completely.

**Comment by janpio** [comment]
> I deployed this project successfully, but it was deployed without `prisma generate` command and still had 50 MB zipped size.

Do you know what changed since then? You said "it was deployed without `prisma generate` command" - what does that mean exactly?

**Comment by NikitaHarzha** [comment]
It didn't generate prisma schema and I got an error
![image](https://user-images.githubusercontent.com/80341254/115387819-d64dc080-a1e3-11eb-98bd-e6be03881a3f.png)

**Comment by janpio** [comment]
Ok, that looks like an error on AWS Lambda. You should probably generate the Client locally using `prisma generate`, then upload the generated artifact (that you are outputting into `./src/database/client` via your schema file) with the package - and then this error should not appear any more. That did not work?

**Comment by NikitaHarzha** [comment]
@janpio I've attached config in description of the bug

**Comment by janpio** [comment]
Anything special about your `handler: src/lambdas/test.handler` file? My `sls package` is currently failing around the step. If you could post the exact path and maybe even content might help.

**Comment by NikitaHarzha** [comment]
@janpio Resources file just contains CORS settings

## Resources file
```
Resources:
  GatewayResponseDefault4XX:
    Type: 'AWS::ApiGateway::GatewayResponse'
    Properties:
      ResponseParameters:
        gatewayresponse.header.Access-Control-Allow-Origin: "'*'"
        gatewayresponse.header.Access-Control-Allow-Headers: "'*'"
      ResponseType: DEFAULT_4XX
      RestApiId:
        Ref: 'ApiGatewayRestApi'
```
`src/lambdas/test.handler` is an exact path to the lambda in my project
Lambda function just has test of the setup of Prisma, just one test request

## Test lambda function
```
import { apiGatewayResponse } from '@/middlewares/apiGatewayResponse';
import { APIGatewayEvent, DefaultEvent, DefaultResponse } from '@/types/aws';
import { APIGatewayResponse } from '@/utils/aws';
import middy from 'middy';
import { v4 as uuidv4 } from 'uuid';
import { jsonBodyParser } from 'middy/middlewares';
import { PrismaClient } from '@prisma/client';

const prisma = new PrismaClient();

const rawHandler = async (
    event: APIGatewayEvent<DefaultEvent>,
): Promise<APIGatewayResponse<DefaultResponse>> => {
    const newUser = await prisma.user.create({
        data: {
            id: uuidv4(),
            name: 'Top',
            email: 'nikita@labda.direct',
            role: 'FREELANCER',
            createdAt: new Date(),
        },
    });

    console.log('New user', newUser);

    return { message: 'Success' };
};

export const handler = middy(rawHandler)
    .use(jsonBodyParser())
    .use(apiGatewayResponse<APIGatewayEvent<DefaultEvent>,
    APIGatewayResponse<DefaultResponse>>());
```
But it has my custom types for the event and response, you will need to use similar types from `aws-sdk` to make lambda work as expected

**Comment by janpio** [comment]
Could not get the `serverless-webpack` plugin to work yet (some strange `TypeError [ERR_INVALID_ARG_TYPE]: The "path" argument must be of type string. Received undefined` error) but I can already reproduce a far to big archive with that plugin disabled - something is strange here and different to our serverless Framework end to end test. Will keep looking into this.

**Comment by NikitaHarzha** [comment]
Did you specify path for the lambda function? It seems like nothing was provided there.

**Comment by janpio** [comment]
When and where would I specify that path?

**Comment by NikitaHarzha** [comment]
```
TestFunction:
    handler: src/lambdas/test.handler
    name: ${self:custom.servicePrefix}-test
    events:
      - http:
          path: test
          method: get
          cors: true
```

It has to be specified like this
`test` is the path for lambda function

**Comment by janpio** [comment]
Ah yes, I copy pasted your configuration from above - that is what is confusing me.

But I'll continue the investigation into the size first before trying to fix this now.

**Comment by AndriiSherman** [comment]
@janpio 

> Some files need to be included of course, otherwise Prisma can not work - the node_modules/.prisma/client/query-engine-rhel-* file is indeed required and that takes ~50MB unpacked. We are working on getting this file smaller over time.

Is there any way not to have such big dependency size? It's so much for lambda (20% of lambda size limit)
I don't think, that it's good for any project

**Comment by janpio** [comment]
See the "We are working on getting this file smaller over time." part of my message 😄 
But: Can you maybe open a feature request about this so we can track it separately? Your reasoning is also enough as the description of the issue 🙇‍♂️

**Comment by AndriiSherman** [comment]
@janpio sure, no problems!

**Comment by janpio** [comment]
Small side note: 
As I was investigating this using a Windows computer, I actually uncovered another bug that affects only Windows and creates much bigger archives there: https://github.com/prisma/prisma/issues/6670

(Actual investigation still ongoing)

**Comment by janpio** [comment]
And another side note, this thing seems to be deep:
As I was investigating the macOS archives from our end to end test runs (which include a Serverless Framework tests) I noticed that the archive was much smaller than on Windows, but still had a few funny files that I could not place why they were included. Turns out that might be another bug causing a 10MB unpacked (2MB packed) size increase that could be avoided: https://github.com/prisma/prisma/issues/6671

(Actual investigation still ongoing)

**Comment by janpio** [comment]
> It didn't generate prisma schema and I got an error
> ![image](https://user-images.githubusercontent.com/80341254/115387819-d64dc080-a1e3-11eb-98bd-e6be03881a3f.png)

Ok, I think I figured out at least why _that_ is happening:
1. You set `output = "./src/database/client"` in the `generator` block of the schema. So when you run `prisma generate` the client is generated to that folder.
2. In your handler file you import `import { PrismaClient } from '@prisma/client';` which imports from the usual Client location, where there is only a default PrismaClient that gives you this error message.

If you update that import, or remove the `output` property in the `generator`, this should work together much better.

This does not explain the too big binary though. Still looking into this.

**Comment by janpio** [comment]
Ok, I think I got _something_:

It seems Serverless Framework has a bug/regression in the packaging behavior (exclusion of DevDependencies) when using Node 15. See the file sizes in an absolutely minimal project with Prisma CLI and Prisma Client (no binaryTargets - so would not work on Lambda) for the two node versions:

- Node 14 is around 20MB https://github.com/janpio/repro6644_3/actions/runs/768797405
- Node 15 is around 80MB https://github.com/janpio/repro6644_3/actions/runs/768802907

This also explains why I had terrible results with your reproduction, and all my tries in between - I was using Node 15.12.0.

---

Problem: Above you reported you are using "Node 12" @NikitaHarzha. What version exactly? Maybe the problem also exists in new Node 12 versions?

To simplify the problem understanding, it would probably be best if you could use my reproduction at https://github.com/janpio/repro6644_3. Please check it out and run the following commands:
```
yarn
yarn prisma generate
yarn sls package
ls -l .serverless
```
How big is the .zip file for you?

**Comment by janpio** [comment]
I tried to apply this Node version knowledge to the reproduction project that I created based on your information earlier, but could see no change. Also removing the `prisma` and `@prisma/client` dependencies did not create a considerable change in the file size of the generated archive.

Would it be possible that Serverless Framework has a problem with your project setup in general @NikitaHarzha, possibly since a recent version? Do you have a lockfile of when this used to work for your project (without Prisma or whatever working state it was in before)?

**Comment by AndriiSherman** [comment]
@janpio just lost track a little bit for current config

> package:
  patterns:
    - '!node_modules/.prisma/client/query-engine-*'
    - 'node_modules/.prisma/client/query-engine-rhel-*'
    
Seems like it didn't help
Still have all `query-engine-*` files inside .zip package uploading to lambda

**Comment by allanitis** [comment]
Have also tried the following config for webpack and the additional files are included. 

```
webpack:
    webpackConfig: webpack.config.js
    includeModules:
      forceExclude:
        - aws-sdk
    packager: npm
    excludeFiles:
      - node_modules/.prisma/client/query-engine-darwin
    packagerOptions:
      scripts:
        - 'prisma generate'
```

I have tested with and without the `prisma generate` script. And the rouge engines are still included in both situations. 

`query-engine-darwin` is the engine for Mac OS. I am testing the bundle locally on MacOS and on the pipeline runner, which uses Debian.

**Comment by NikitaHarzha** [comment]
I've tried all possible exclude dependency setup. 
Tried to use `forceExclude`, `excludeFiles`, `excludeRegex` all of these are `serverless-webpack` options, but any of this work for me.

**Comment by janpio** [comment]
@AndriiSherman That is weird, as you are explicitly asking Serverless Framework to exclude that path, right? Can you please open a new issue? Then we can debug this in detail.

@allanitis The explicit `excludeFiles` is ignored, or are there _other_ engines in the archive? Optimally also create a new issue and we can debug this individually.

@NikitaHarzha In your project I am observing Serverless Framewor including _all_ devDependencies, even when Prisma and Prisma Client are not installed. Can you confirm this behavior? The bundled file via `sls package` should in theory only include dependencies and their dependencies in the `node_modules` folder.

---

I will come up with a more general `package.patterns` entry that should include all files from Prisma CLI and Prisma Client later, to work around Serverless Framework _somehow_, _sometimes_ not excluding devDependencies. But this will unfortunately only help for Prisma itself. I fear you are only noticing the Prisma inclusion because of the file size (which of course makes it most pressing and relevant), but it _looks_ to me that Serverless Framework is failing to exclude unneeded dependencies in general somehow.

**Comment by rubenkaiser** [comment]
> Hm, that is weird @NikitaHarzha - sorry about that.
> 
> Some files _need_ to be included of course, otherwise Prisma can not work - the `node_modules/.prisma/client/query-engine-rhel-*` file is indeed required and that takes ~50MB unpacked. We are working on getting this file smaller over time.
> 
> 1. Did you deploy this project before successfully?
> 2. Can you use `serverless package` and then `less .sererless/####.zip` to look into the generated archive, what is taking space in there?
>    
>    * Feel free to also send me the zip file privately to [piotrowski@prisma.io](mailto:piotrowski@prisma.io) if you want me to take a look. I know what to look for now.
> 3. Is your project expected to use >200MB for the rest of the code?

Deploying with Vercel we also experience this problem. Since vercel does the deployment only the query-engine for AWS is in the output by default. So no steps are required to exclude or remove others. However going from version 2.19.x to 2.20.1 the size of query-engine-rhel-openssl-1.0.x has increased a lot. It has gone from about 25MB to 45MB causing our deployments to fail because of the 50MB limit upload limit of AWS lambda.

If others are experiencing this the solutions would be to use 2.19 for now, as they are working on getting the size down as denoted by @janpio

**Comment by janpio** [comment]
> It has gone from about 25MB to 45MB causing our deployments to fail because of the 50MB limit upload limit of AWS lambda.

Note that the file increase of the binary is unpacked, while the 50MB limit is for the zip file. But yes, in general this is a valid concern. You thankfully opened an additional issue at https://github.com/prisma/prisma/issues/6746, so let's debug your case there. Usually the binary size it not that big of a problem for projects.

**Comment by carlosherrera1113** [comment]
If this helps shed any light on the issue, I'm having size issues because serverless will not remove the binaries in node_modules/.prisma/client via: 
``` 
 patterns:
    - "!node_modules/.prisma/client/query-engine-*"
    - "node_modules/.prisma/client/query-engine-rhel-*"
 ```

**Comment by janpio** [comment]
Which binaries are still included for you @carlosherrera1113? Optimally open a new issue so we can debug your issue individually.

**Comment by styxlab** [comment]
I am experiencing the same as @rubenkaiser on Vercel. Pinning prisma to `2.19.0` solves this issue, but I am looking for ways to reduce the prisma footprint and get ideas on how to do that when deploying to Vercel.
```
node_modules/@prisma/engines  113 MB          39.6 MB
node_modules/.prisma/client  46.2 MB          16.1 MB
node_modules/prisma/build     5.3 MB          1.09 MB
node_modules/@prisma/client  3.93 MB           823 kB
```
Any help is appreciated.

**Comment by Jolg42** [comment]
@styxlab Could you create a new issue with your package.json dependencies?

Note: `node_modules/@prisma/engines` should normally not be uploaded, this dependency is only for the CLI and the CLI should normally be added as a `devDependency` only, maybe you have it has a `dependency` and putting it as a `devDependency` already fixes your issue.

**Comment by styxlab** [comment]
@Jolg42 Thanks so much for looking into this. I opened new case #7016 as suggested. I am not sure how I can control `node_modules/@prisma/engines` as I only have the following prisma dependencies in `packages.json` (showing the currently working case). When upgrading those two packages to > "2.19.0" I run into the AWS lambda size limits on Vercel.

```
  "dependencies": {
   ...
    "@prisma/client": "2.19.0",
  },
  "devDependencies": {
    ...
    "prisma": "2.19.0"
  }
```
There is more detailed info on the new case.

**Comment by jsw** [comment]
I've run into the same issue, with the `package/patterns` config appearing to not exclude the darwin engine.

```
functions:
  api:
    handler: 'dist/serverless.handler'
    events:
      - httpApi:
          path: /
          method: '*'
      - httpApi:
          path: /{proxy+}
          method: '*'
    package:
      patterns:
        - '!node_modules/.prisma/client/query-engine-*'
        - 'node_modules/.prisma/client/query-engine-rhel-*'
```

Run `serverless package`

```
unzip -l .serverless/wb2.zip | grep query-engine
 37170408  01-01-1980 00:00   node_modules/.prisma/client/query-engine-darwin
 45213280  01-01-1980 00:00   node_modules/.prisma/client/query-engine-rhel-openssl-1.0.x
 37170408  01-01-1980 00:00   node_modules/@prisma/engines/query-engine-darwin
 37170408  01-01-1980 00:00   node_modules/prisma/query-engine-darwin
 45213280  01-01-1980 00:00   node_modules/prisma/query-engine-rhel-openssl-1.0.x
```

**Comment by jsw** [comment]
@janpio I created a minimum repro at https://github.com/jsw/prisma-issue-6644

Also, I fixed part of the issue by moving the `packages.patterns` from function scope to file scope, though both are supposed to work if I'm correctly following the serverless docs. Take a look at my `serverless.yml`. And `@prisma/engines` and `prisma` are still included unless your workaround above is applied.

**Comment by janpio** [comment]
> @janpio Could this suggestion solve the Vercel cold start issues #7303 ?

I am not sure, best you just try it.

> For those using serverless platform, since upgrading to prisma 3, I have updated my serverless package patterns to the following, which seems to cover everything

Thanks @jsw - do you want to send a PR to these docs pages?

**Comment by saifxx** [comment]
As my development environment is Windows 10. After using @jsw solution, my package size is still 62MB.  I have updated my serverless package patterns to the following, and now my package size is 19MB.

`
    - '!node_modules/.prisma/client/libquery_engine-*'
    - 'node_modules/.prisma/client/libquery_engine-rhel-*'
    - '!node_modules/prisma/libquery_engine-*'
    - '!node_modules/@prisma/engines/**'

    # EXCLUDE query_engine-windows.dll.node
    - '!node_modules/.prisma/client/query_engine-*'
    # EXCLUDE .cache folder
    - '!node_modules/.cache/**'
    # EXCLUDE .bin folder
    - '!node_modules/.bin/**'
`

**Comment by C5H8NNaO4** [comment]
I'm having the same issue on windows 10, but this package pattern doesn't change anything for me.

**Comment by Jolg42** [comment]
@C5H8NNaO4 Could you open [a new bug issue her](https://github.com/prisma/prisma/issues/new?assignees=&labels=kind%2Fbug&template=bug_report.yml), this issue is old and filling a new issue with your details will be useful, it might be a bit different in your case.

**Comment by janpio** [comment]
Today's release 4.8.0 has _significantly_ (average 50%, on some platforms even more) decreased the size of our engine files:
> **Improved serverless experience — smaller engines size**
>
> In this release, we have decreased the size of our engine files by an average of 50%. The size of the Query Engine used on Debian, with OpenSSL 3.0.x, for example, went from 39MB to 14MB. We will also remove some smaller engines to decrease the total size in future versions.
>
> Additionally, we have started optimizing how the Prisma schema is loaded in Prisma Client. You should notice a considerable improvement when executing the first query if you're working with a bigger schema with many models and relations.
>
> We will continue investing in this direction in the next releases and further improve the experience with Prisma and serverless environments.

https://github.com/prisma/prisma/releases/tag/4.8.0

We know we still have duplicated engine files in multiple locations, and that we are including too many of our engine files - but this was the obvious place to start and we are happy how significantly we could improve the situation already. Expect more updates like this in the coming releases from us.

**Comment by janpio** [comment]
As written before, we have reduced the size of our engines a lot and as far as we know, this solved this problem for most our user and their use cases. Hence I am closing this issue for now.

If you and your project are still having problems because of the remaining size of our engines, please open a new issue and describe your situation - we are happy to take a look and hopefully fix that. Thanks!

**Comment by bjj951** [comment]
This problem raised it head again for me after performing an npm update, either serverless or npm is now additionally generating under node_modules/.cache/prisma which were getting included in the zip and exceeding the 250mb limit. I updated my serverless.yml to also exclude this directory as follows:

      patterns:
        - '!node_modules/.prisma/client/libquery_engine-*'
        - 'node_modules/.prisma/client/libquery_engine-rhel-*'
        - '!node_modules/prisma/libquery_engine-*'
        - '!node_modules/@prisma/engines/**'
        - '!node_modules/.cache/prisma/**'

my global package versions are as follows:
npm@9.1.2
serverless@3.22.0
tsc@2.0.4
 typescript@4.8.2

and prisma@5.11.0

**Comment by Jolg42** [comment]
@bjj951 please open a new issue or a discussion as this issue is closed since almost a year. So we can process this.
I think this would be best as an issue in our docs repository: https://github.com/prisma/docs

From what I understand, you are on Windows? I think this directory `node_modules/.cache/prisma/` is what you are talking about? It is only present for Windows users.

The related pages in the docs should be:
- https://www.prisma.io/docs/orm/prisma-client/deployment/serverless/deploy-to-aws-lambda#deploy-only-the-required-files
- https://www.prisma.io/docs/orm/more/upgrade-guides/upgrading-versions/upgrading-to-prisma-5#example-with-the-serverless-framework

**Comment by bjj951** [comment]
@Jolg42, i created a patch for the docs here to add that missing pattern for Windows. 
https://github.com/prisma/docs/compare/main...bjj951:docs:patch-1

**Comment by Jolg42** [comment]
Thanks, I opened a PR from your patch at https://github.com/prisma/docs/pull/5726
